src/ucp/wireup/wireup_cm.c

```diff
@@ -439,23 +463,57 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
     UCS_STATIC_ASSERT(ucs_static_array_size(ep_init_flags_prio) > 0);
 
     ucp_wireup_eps_pending_extract(ep, &tmp_pending_queue);
-    for (i = 0; i < ucs_static_array_size(ep_init_flags_prio); ++i) {
-        ep_init_flags = ep_init_flags_prio[i];
 
-        /* Cleanup the previously created UCP EP. The one that was created on
-         * the previous call to this client's resolve_cb */
-        ucp_wireup_cm_ep_cleanup(ep);
+    /* Cleanup the previously created UCP EP. The one that was created on
+     * the previous call to this client's resolve_cb */
+    ucp_wireup_cm_ep_cleanup(ep);
 
-        status = ucp_cm_ep_client_initial_config_get(ep, ep_init_flags,
+    for (i = 0; i < ucs_static_array_size(ep_init_flags_prio); ++i) {
+        status = ucp_cm_ep_client_initial_config_get(ep, ep_init_flags_prio[i],
                                         &cm_wireup_ep->cm_resolve_tl_bitmap,
                                         &key);
         if (status != UCS_OK) {
             goto try_fallback;
         }
 
+        if (i < (ucs_static_array_size(ep_init_flags_prio) - 1)) {
+            /* Can use another variant of endpoint's initialization flags for
+             * fallback */
+            fallback_log_level = UCS_LOG_LEVEL_DEBUG;
+        } else if (ucp_cm_client_get_next_cm_idx(ep) != UCP_NULL_RESOURCE) {
+            /* Can use another CM for fallback */
+            fallback_log_level = UCS_LOG_LEVEL_DIAG;
+        } else {
+            /* No options for fallback */
+            fallback_log_level = UCS_LOG_LEVEL_ERROR;
+        }
+
+        ucp_ep_get_tl_bitmap(&key, &tl_bitmap);
+        ep_pack_flags = ucp_wireup_cm_address_pack_flags(worker, cm_wireup_ep,
+                                                         ep_init_flags_prio[i]);
+        status        = ucp_address_length(worker, &key, &tl_bitmap,
+                            ep_pack_flags, context->config.ext.worker_addr_version,
+                            &data_length);
+        if (status != UCS_OK) {
+            goto err;
+        }
+
+        status = ucp_wireup_cm_priv_space_check(worker,
+                     ep->ext->cm_idx, data_length,
+                     context->config.ext.sa_client_min_hdr_version,
+                     fallback_log_level);
+        if ((status != UCS_OK) && (status != UCS_ERR_BUFFER_TOO_SMALL)) {
+            goto err;
+        }
+
+        if ((status == UCS_ERR_BUFFER_TOO_SMALL) &&
+            (i < ucs_static_array_size(ep_init_flags_prio) - 1)) {
```

Comment:
why need to check `(i < ucs_static_array_size(ep_init_flags_prio) - 1)` here?

Reply:
according to the code, when buffer is too small, but `i == ucs_static_array_size(ep_init_flags_prio) - 1)`, we will continue to create uct_eps. but we shouldn't.

Reply:
Done. Simplify the logic without extracting & replying the pending request when CM buffer is too small.